frame/transaction-payment/src/lib.rs

```diff
@@ -214,33 +264,41 @@ impl<T: Trait + Send + Sync> SignedExtension for ChargeTransactionPayment<T>
 		info: Self::DispatchInfo,
 		len: usize,
 	) -> TransactionValidity {
-		// pay any fees.
 		let tip = self.0;
 		let fee = Self::compute_fee(len as u32, info, tip);
-		// Only mess with balances if fee is not zero.
-		if !fee.is_zero() {
-			let imbalance = match T::Currency::withdraw(
-				who,
-				fee,
-				if tip.is_zero() {
-					WithdrawReason::TransactionPayment.into()
-				} else {
-					WithdrawReason::TransactionPayment | WithdrawReason::Tip
-				},
-				ExistenceRequirement::KeepAlive,
-			) {
-				Ok(imbalance) => imbalance,
-				Err(_) => return InvalidTransaction::Payment.into(),
-			};
-			T::OnTransactionPayment::on_unbalanced(imbalance);
-		}
+		let _ = Self::withdraw_fee(who, tip, fee)
+			.map_err(|()| TransactionValidityError::from(InvalidTransaction::Payment))?;
 
 		let mut r = ValidTransaction::default();
 		// NOTE: we probably want to maximize the _fee (of any type) per weight unit_ here, which
 		// will be a bit more than setting the priority to tip. For now, this is enough.
 		r.priority = fee.saturated_into::<TransactionPriority>();
 		Ok(r)
 	}
+
+	fn post_dispatch(
+		pre: Option<PendingRefund<T::AccountId, NegativeImbalanceOf<T>>>,
+		_info: Self::DispatchInfo,
+		_len: usize,
+	) {
+		let (transactor, imbalance) = match pre {
+			Some(PendingRefund {
+				transactor,
+				imbalance,
+			}) => (transactor, imbalance),
+			_ => return,
+		};
+
+		let actual_weight_fee = <frame_system::Module<T>>::spent_weight()
+			.map(|spent| T::WeightToFee::convert(spent))
+			.unwrap_or_default();
+		let refund = imbalance.peek().saturating_sub(actual_weight_fee);
+
+		let refund_imbalance = T::Currency::deposit_creating(&transactor, refund);
+		if let Ok(imbalance) = imbalance.offset(refund_imbalance) {
+			T::OnTransactionPayment::on_unbalanced(imbalance);
```

Comment:
I am not sure what you mean here because panic is still a pretty-much a game over scenario atm since in case of a panic all changes are reverted, including `pre_dispatch`. There are some mitigations planned, but I wouldn't say that they are on the horizon yet.

Reply:
yeah totally my misunderstanding ✋